ci/swagger docs

a20688392 opened the pull request:
Setup integrate SwaggerHub with GitHub, generate swagger file on SwaggerHub.

## Use SwaggerHub CLI
### [SwaggerHub CLI link](https://github.com/SmartBear/swaggerhub-cli/tree/main#swaggerhub-cli)
- Set Actions secrets and variables
  - SWAGGERHUB_API_KEY：use API Key for swaggerhub authentication.

## Execution conditions
### Version
Version according to package.json
### In PR
- main
  because only one version will be added at a time, so swaggerhub-cli use create api.
  but need to upgrade the version in package.json.
  https://app.swaggerhub.com/apis/Cophr/main
- others
  will use the version of package.json for this branch, so swaggerhub-cli use update api.
  https://app.swaggerhub.com/apis/Cophr/test

a20688392 committed:
ci: generate swagger file

a20688392 committed:
ci: upload file on Swaggerhub

moontai0724 commented on .github/workflows/ci.yml:
```diff
@@ -67,3 +67,57 @@ jobs:
+      - name: Upload on Swaggerhub - Cosphr_main
+        if: github.ref == 'refs/heads/main'
+        run: |
+          swaggerhub api:create "a20688392/Cosphr/${{ env.APP_SWAGGER_Version }}" --setdefault --file=swagger-docs.json --visibility=public
```
I would prefer to create a Cophr official account to put instead of using your personal account.

a20688392 replied:
Switched to official account
[fix: organization name correction](https://github.com/Cophr/backend/pull/9/commits/e7c99c3c7f866bbb6fd3df1815d316007fac8ea2)

moontai0724 commented on .github/workflows/ci.yml:
```diff
@@ -67,3 +67,57 @@ jobs:
+      - name: Upload on Swaggerhub - Cosphr_main
+        if: github.ref == 'refs/heads/main'
+        run: |
+          swaggerhub api:create "a20688392/Cosphr/${{ env.APP_SWAGGER_Version }}" --setdefault --file=swagger-docs.json --visibility=public
```
Our name is `Cophr` instead of `Cosphr`.

a20688392 replied:
Corrected but switched to official account.
[fix: organization name correction](https://github.com/Cophr/backend/pull/9/commits/e7c99c3c7f866bbb6fd3df1815d316007fac8ea2)

moontai0724 reviewed (changes requested):
That is a big mistake. LOL

a20688392 committed:
fix: organization name correction

moontai0724 commented on package.json:
```diff
@@ -8,6 +8,7 @@
   "scripts": {
     "prebuild": "rimraf dist",
     "build": "nest build",
+    "generate-swagger": "node ./dist/swagger/generate-swagger-file.js",
```
This requires to build project into js first, may be able to add the build process into the command or use `ts-node` to run it directly

a20688392 replied:
I don't think it's necessary, it solves our needs, but can consider in the future.

moontai0724 commented on .github/workflows/ci.yml:
```diff
@@ -67,3 +67,57 @@ jobs:
+      - name: Upload on Swaggerhub - Cophr_develop
+        if: github.ref != 'refs/heads/main'
+        run: |
+          swaggerhub api:update "Cophr/test/${{ env.APP_SWAGGER_Version }}" --setdefault --file=swagger-docs.json --visibility=public
```
Use app version may easily result in conflict between PRs, may use commit hash or use other that may not easily to conflict to identify the version.

a20688392 replied:
I think use commit hash as the version is a good idea！

a20688392 replied:
In this commit is solved version by commit hash.
[refactor: swagger version by commit hash](https://github.com/Cophr/backend/pull/9/commits/aa357cea5759e832e4863f47b6d57075f0fb0c0f)
But in order to achieve automatic, use create instead of update.
[fix: not use update is use create](https://github.com/Cophr/backend/pull/9/commits/7d487a2219c18f65f66c0f7d91075d9da230554d)

moontai0724 commented on .github/workflows/ci.yml:
```diff
@@ -67,3 +67,57 @@ jobs:
+        env:
+          APP_SWAGGER_Version: ${{ steps.package-version.outputs.version }}
+
+      - name: Upload on Swaggerhub - Cophr_develop
```
recommend to align naming with below (`test`)

a20688392 replied:
I revised it in this commit. [264c533](https://github.com/Cophr/backend/pull/9/commits/264c533aa3e1dd913b184803ec32a21da0b1e7c2)
You are right. Thank you review.

moontai0724 reviewed (changes requested):
Comments above

a20688392 committed:
refactor: revise description

a20688392 committed:
refactor: swagger version by commit hash

a20688392 committed:
fix: use SwaggerHub API create instead of update

why need? In order to achieve automatic, swagger docs change will follow the commit hash.

create will generate new swagger docs.
update will update specified version information.

moontai0724 reviewed (approved):
LGTM! but it would be better if you reply the swagger url to the PR
